tonic/src/codec/compression.rs

```diff
@@ -0,0 +1,161 @@
+        } else {
+            None
+        }
+    }
+}
+
+/// The compression encodings Tonic supports.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum CompressionEncoding {
+    #[allow(missing_docs)]
+    Gzip,
+}
+
+impl CompressionEncoding {
+    /// Based on the `grpc-accept-encoding` header, pick an encoding to use.
+    pub(crate) fn from_accept_encoding_header(
+        map: &http::HeaderMap,
+        enabled_encodings: EnabledCompressionEncodings,
+    ) -> Option<Self> {
+        let header_value = map.get(ACCEPT_ENCODING_HEADER)?;
+        let header_value_str = header_value.to_str().ok()?;
+
+        let EnabledCompressionEncodings { gzip } = enabled_encodings;
+
+        split_by_comma(header_value_str).find_map(|value| match value {
+            "gzip" if gzip => Some(CompressionEncoding::Gzip),
+            _ => None,
+        })
+    }
+
+    /// Get the value of `grpc-encoding` header. Returns an error if the encoding isn't supported.
+    pub(crate) fn from_encoding_header(
+        map: &http::HeaderMap,
+        enabled_encodings: EnabledCompressionEncodings,
+    ) -> Result<Option<Self>, Status> {
+        let header_value = if let Some(value) = map.get(ENCODING_HEADER) {
+            value
+        } else {
+            return Ok(None);
+        };
+
+        let header_value_str = if let Ok(value) = header_value.to_str() {
+            value
+        } else {
+            return Ok(None);
+        };
+
+        let EnabledCompressionEncodings { gzip } = enabled_encodings;
+
+        match header_value_str {
+            "gzip" if gzip => Ok(Some(CompressionEncoding::Gzip)),
+            other => Err(Status::unimplemented(format!(
+                "Content is compressed with `{}` which isn't supported",
+                other
+            ))),
+        }
+    }
+
+    pub(crate) fn into_header_value(self) -> http::HeaderValue {
+        match self {
+            CompressionEncoding::Gzip => http::HeaderValue::from_static("gzip"),
+        }
+    }
+}
+
+impl fmt::Display for CompressionEncoding {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CompressionEncoding::Gzip => write!(f, "gzip"),
+        }
+    }
+}
+
+fn split_by_comma(s: &str) -> impl Iterator<Item = &str> {
+    s.trim().split(',').map(|s| s.trim())
+}
+
+/// Compress `len` bytes from `decompressed_buf` into `out_buf`.
+pub(crate) fn compress(
+    encoding: CompressionEncoding,
+    decompressed_buf: &mut BytesMut,
+    out_buf: &mut BytesMut,
+    len: usize,
+) -> Result<(), std::io::Error> {
+    let capacity = ((len / BUFFER_SIZE) + 1) * BUFFER_SIZE;
+    out_buf.reserve(capacity);
+
+    match encoding {
+        CompressionEncoding::Gzip => {
+            let mut gzip_encoder = GzEncoder::new(
+                &decompressed_buf[0..len],
+                // FIXME: support customizing the compression level
+                flate2::Compression::new(6),
```

Comment:
It would be nice to support customizing this. I'll look into that when this has been merged while I also look into adding support for more compression encodings.